node_package/src/getReactServerComponent.client.ts

```diff
@@ -104,41 +131,48 @@ const createFromPreloadedPayloads = (payloads: string[]) => {
+ * on server has different parameters than the function that gets them on client. The environment
+ * dependent parameters (domNodeId, railsContext) are passed from the `wrapServerComponentRenderer`
+ * function, while the environment agnostic parameters (componentName, componentProps, enforceRefetch)
+ * are passed from the RSCProvider which is environment agnostic.
+ *
+ * The returned function:
+ * 1. Checks for embedded RSC payloads in window.REACT_ON_RAILS_RSC_PAYLOADS using the domNodeId
  * 2. If found, uses the embedded payload to avoid an HTTP request
  * 3. If not found (during client navigation or dynamic rendering), fetches via HTTP
  * 4. Processes the RSC payload into React elements
  *
  * The embedded payload approach ensures optimal performance during initial page load,
  * while the HTTP fallback enables dynamic rendering after navigation.
  *
+ * @param domNodeId - The DOM node ID to create a unique key for the RSC payload store
+ * @param railsContext - Context for the current request, shared across all components
+ * @returns A function that accepts RSC parameters and returns a Promise resolving to the rendered React element
+ *
+ * The returned function accepts:
  * @param componentName - Name of the server component to render
  * @param componentProps - Props to pass to the server component
- * @param railsContext - Context for the current request
  * @param enforceRefetch - Whether to enforce a refetch of the component
- * @returns A Promise resolving to the rendered React element
  *
  * @important This is an internal function. End users should not use this directly.
  * Instead, use the useRSC hook which provides getComponent and refetchComponent functions
  * for fetching or retrieving cached server components. For rendering server components,
  * consider using RSCRoute component which handles the rendering logic automatically.
  */
-const getReactServerComponent = ({
-  componentName,
-  componentProps,
-  railsContext,
-  enforceRefetch = false,
-}: ClientGetReactServerComponentProps) => {
-  assertRailsContextWithComponentSpecificMetadata(railsContext);
-  const componentKey = createRSCPayloadKey(componentName, componentProps, railsContext);
-  const payloads = window.REACT_ON_RAILS_RSC_PAYLOADS?.[componentKey];
-  if (!enforceRefetch && payloads) {
-    return createFromPreloadedPayloads(payloads);
-  }
-  return fetchRSC({ componentName, componentProps, railsContext });
-};
+const getReactServerComponent =
+  (domNodeId: string, railsContext: RailsContext) =>
+  ({ componentName, componentProps, enforceRefetch = false }: ClientGetReactServerComponentProps) => {
+    const componentCacheKey = createServerComponentCacheKey(componentName, componentProps);
+
+    const rscPayloadKey = `${componentCacheKey}-${domNodeId}`;
+    const payloads = window.REACT_ON_RAILS_RSC_PAYLOADS?.[rscPayloadKey];
+    if (!enforceRefetch && payloads) {
+      return createFromPreloadedPayloads(payloads);
+    }
```

Comment:
```suggestion
    if (!enforceRefetch) {
      const componentCacheKey = createServerComponentCacheKey(componentName, componentProps);
      const rscPayloadKey = `${componentCacheKey}-${domNodeId}`;
      const payloads = window.REACT_ON_RAILS_RSC_PAYLOADS?.[rscPayloadKey];
      if (payloads) {
        return createFromPreloadedPayloads(payloads);
      }
    }
```
since `payloads` are only needed in that case. Or even `if (!enforceRefetch && window.REACT_ON_RAILS_RSC_PAYLOADS)` (and remove `?.`).